docs/proposal/20230317-irsa-support-for-awscluster.md

```diff
@@ -0,0 +1,152 @@
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+- [CAPA](https://cluster-api.sigs.k8s.io/reference/glossary.html#capa) - Cluster API Provider AWS. 
+- [CAPI](https://github.com/kubernetes-sigs/cluster-api) - Cluster API.
+- [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) - IAM Roles for Service Accounts
+- [pod-identity-webhook](https://github.com/aws/amazon-eks-pod-identity-webhook) - Pod Identity Webhook Repo
+
+## Summary
+The IAM Roles for Service Accounts take the access control enabled by IAM and bridge the gap to Kubernetes by adding roles-based access to service accounts. CAPA users of self-managed clusters can now give granular roles-based access to the AWS API at a pod level.      
+
+## Motivation
+This functionality is currently built into EKS, with a simple boolean in the AWSManagedCluster API called `AssociateOIDCProvider` CAPA will build an IAM OIDC provider for the cluster and create a trust policy template in a config map to be used for created IAM Roles. Self-managed clusters can use IRSA but require additional manual steps already done in Managed Clusters, including patching kube-api-server, creating an OIDC provider and deploying the `pod-identity-webhook`, which is documented in their [self-hosted setup](https://github.com/aws/amazon-eks-pod-identity-webhook/blob/master/SELF_HOSTED_SETUP.md) walkthrough but with CAPA style ingredients like using the management cluster, kubeadm config modification and the built-in serving certs' OpenID Configuration API endpoints.
+
+The pieces to IRSA are easily created with the existing access for CAPA. By adding `AssociateOIDCProvider` to `AWSCluster` we can kick off a reconciliation process to generate all pieces necessary to utilize IRSA in your self-managed cluster. 
+
+### Goals
+
+1. On cluster creation, add all components to self-managed clusters to use IAM Roles for Service Accounts.
+2. On cluster deletion, remove all external dependencies from the AWS account.
+
+### Non-Goals/Future Work
+Migrate all IAM work for Managed cluster to the IAM service. 
+
+## Proposal
+- Create a boolean on `AWSCluster` called `AssociateOIDCProvider` to match the `AWSManagedCluster` API and have a default value of `false`.
+- Migrate the status types for `OIDCProvider` out of the experimental EKS APIs and into the v1beta2 APIs. 
+- Build an IAM cloud service and add a reconciler to work to persist all components required for IRSA; the logic is as follows.
+  1. Create a self-signed issuer for the workload cluster namespace to be used to make the pod identity webhook serving cert.
+  2. Generate the patch file and update kubeadm configs to write the patch to disk for the control plane nodes.
+  3. Create the Identity Provider in IAM pointed to the S3 bucket.
```

Comment:
This is the first mention of a bucket. What is in this bucket and why is it needed?

Reply:
There needs to be a public hosted jwks and open mid config for oidc to work. storing these in s3 is a simple way to have have two files public for sts token auth

Reply:
That is fine, just mention that in the doc as well as a sidenote or a footnote :)

Reply:
I still don't see any explanation what the S3 bucket will contain. Just write down something like, and S3 bucket exists that contains bla.

Reply:
Hey @luthermonson. Can you add this note please before this part? I can approve it then. :)

Reply:
@Skarlso Added more details and sample s3 directory structure.

Reply:
Excellent, thanks!